### ISSUE TITLE
No fonts in project manager if using custom theme.

### ISSUE THREAD

**Issue by autofool** [godotengine/godot#48477]
**Godot version:**
3.3-stable

**Issue description:**
If I set custom fonts in editor theme resource, fonts won't show up in project manager. It is not even using any kind of fallback fonts.

**Steps to reproduce:**
Set theme file given below as custom theme in godot editor. Then press `Ctlr+Shift+q` to go back to project manager and it will not show fonts. Now go back to any project and copy theme in project folder. Edit it in inspector and just remove default font option. Now go back to project manager. Some of the font's come back except for few places. That's because theme is overriding more fonts in sub properties. If you remove all custom fonts. Everything will show properly .

**Minimal reproduction project:**
[material.zip](https://github.com/godotengine/godot/files/6426232/material.zip)

**Comment by JestemStefan** [comment]
This behaviour is documented. 
![obraz](https://user-images.githubusercontent.com/37214990/117128582-e3e77680-ad9d-11eb-9610-7bee2ab8aaa7.png)

The workaround is to apply theme to Control node

**Comment by pycbouh** [comment]
@JestemStefan Hey, which doc is this lifted from? I've actually addressed this propagation problem in one of my PRs so I would need to update the docs if/when it gets merged.

**Comment by JestemStefan** [comment]
@pycbouh https://docs.godotengine.org/en/3.3/tutorials/gui/gui_skinning.html

**Comment by a3p169** [comment]
@JestemStefan I'm sorry but I don't understand. How would applying theme to control node instance would affect Godot editor? Are we on same page? I'm not using theme for any gui scene for game but for editor itself.

**Comment by a3p169** [comment]
![20210505162901_REGION](https://user-images.githubusercontent.com/50569653/117131443-1190e780-adbf-11eb-8dfb-e723ed9d0352.png)
This is the theme resource I'm talking about.
![20210505162956_REGION](https://user-images.githubusercontent.com/50569653/117131501-24a3b780-adbf-11eb-9d4a-0ff86b6e36eb.png)
And that's project manager without any fonts.

**Comment by a3p169** [comment]
I have set default font's in theme resource. So it should use it globally if I have not set them somewhere.

**Comment by pycbouh** [comment]
This does seem like an independent issue, yes.

**Comment by JestemStefan** [comment]
OK. I thought you are talking about custom theme for GUI. IDK, about this one, but maybe they are related? Someone more advanced need to look into it.

**Comment by Calinou** [comment]
The custom editor theme setting has been broken for a while. I'd remove it entirely at this point. The customization options offered in the built-in theme are sufficient for nearly every use case.

**Comment by a3p169** [comment]
@Calinou Removing it would be really sad cause it's a really cool feature! I think custom theme for editor which uses the resource created by the editor is pretty fascinating. Plus, it gives users a prime example of Godot's capability to theme components without manually updating them. Hack we can even put some example on wiki so users can just browse editor with applied theme and check how it works without creating a whole scene with control skinning cause that has its own learning curve. This way new users who are still considering Godot can check out how GUI skinning works without actually learning Godot. And it should be recommended on features page (If custom theme is fixed). 
Another thing is that we spend most of our time using editor and having some personal touch makes it more enjoyable. I personally take at least a whole day to set up any editor I'm going to use to my liking. On top of that default settings are a bit weird. Especially the contrast.
Contrast makes border panels more brighter and that's fine but there is no way to inverse contrast so that primary color is background color which is darkened. I like having my application look seamless with my desktop theme and contrast is really making things ugly.
If I want to achieve flat looks with it, its basically impossible due to the fact that Godot has very congested UI. It's only possible with manually tweaking the theme. Default option lower contrast just makes things harder to distinguish and confusing.
Improved custom theme is not an immediate issue that needs fixing asap but its fix is definitely one that would make day to day use of Godot more likeable. And it's not like custom theme is totally broken. It works pretty well just needs more robust application.

**Comment by Calinou** [comment]
> Contrast makes border panels more brighter and that's fine but there is no way to inverse contrast so that primary color is background color which is darkened. I like having my application look seamless with my desktop theme and contrast is really making things ugly.

That's something that could be exposed either by allowing negative contrast values or adding a new setting 🙂 

I would prefer we improve the built-in theme customizations (to a reasoanble extent).

**Comment by a3p169** [comment]
> Also, the master branch just got a new default editor theme with a flatter, more modern design.
Wut!? Imma head out for a sec.

**Comment by a3p169** [comment]
? Do you mean `tres` file?

**Comment by Calinou** [comment]
> ? Do you mean `tres` file?

Yes, but I deleted my comment as I realized it's in the MRP.

**Comment by pycbouh** [comment]
> It's nice to advocate for the custom theme option, but until now, there is nobody to fix it...

I take personal responsibility to make sure custom theme resources for the editor are in a working order. I see no reason why we wouldn't support a fully custom theme resource, if there are users willing to take time to create one. We use our internal Theme resource in an axiomatic way, so everything should be overridable. A few bugs are not a good reason to just cut it.

I would even argue we don't make our customization options more complicated. I like that Godot's theme can be adjusted in a very simplistic way. For everything more complex, there is (and will be) the Theme editor.

**Comment by Calinou** [comment]
> Isn't option already there?
> ![20210507191219_REGION](https://user-images.githubusercontent.com/50569653/117458874-615ee280-af68-11eb-8ff7-3a197d23101b.png)

`doc_bold` is the font for the editor help (built-in documentation), not the editor in general. `bold` seems to be the option you want here :slightly_smiling_face:

**Comment by a3p169** [comment]
Yah that's what I was talking about. cursor was on doc bold so that popup came up lol

**Comment by Calinou** [comment]
> Umm I think Godot configuration broke. After I compiled and ran Godot from master branch. 3.3 one stopped applying theme to the project manager. I deleted the .config folder in home dir but issue still persists. Any solution to this?

Godot 3 and 4 use different configuration files (see [File paths in Godot projects](https://docs.godotengine.org/en/latest/tutorials/io/data_paths.html) for their location).

**Comment by a3p169** [comment]
OK, I had older version of the theme in backup and it's working in project manager. But the new one doesn't work. I mean it works once you select project and go into editor but the project manager just go backs to default blue theme. Weird. AND this happened after I ran Godot 4. I think it's not an issue for regular user who doesn't run Godot from master.

**Comment by a3p169** [comment]
Wow, theme file is so big I can't make heads or tails of it.

**Comment by Calinou** [comment]
> Wow, theme file is so big I can't make heads or tails of it.

When you create a theme from the editor theme, it contains embedded binary resources. If you save it as a `.tres` file, binary resources are converted to Base64 which is larger and slower to load than an actual binary file. I recommend saving such themes as binary `.res` files instead, so that they load faster and are smaller.

**Comment by a3p169** [comment]
I can confirm that setting fallback font (I tried setting for every instance available in theme resource) doesn't work. What if project manager can't access external resources in theme file before loading the project screen? Is it possible that something like importer has to be available before Godot can load external resource like Image or Font? But that doesn't make sense because it can load project icons.

**Comment by pycbouh** [comment]
So, the problem is not specific to the project manager, it's an issue with Theming in general. And it doesn't even originate in the `Theme` resource. The problem is that both a `DynamicFont` resource and its `DynamicFontData` sub-resource are considered valid even though the external font asset could not be loaded because it doesn't exist at the specified path. I'm not entirely sure if how we can address that. But what this means is that as your theme has a definition for the font the PM is looking for (here being `title` in `EditorFonts`), this definition is used regardless of its actual applicability, and we never fall back to the default font.

Even if we could fix the fallback, there is still the issue of actually loading external resources. There is no project context in the project manager (duh!), so it's unlikely `res://` paths can be resolved, and even if they could, it would likely be the folder with the Godot executable. To solve this properly we need to allow relative paths for sub-resources, see https://github.com/godotengine/godot-proposals/issues/2327.

**Comment by a3p169** [comment]
@pycbouh how does Godot find project icons then?

**Comment by pycbouh** [comment]
By manually adjusting the `res://` path and manually loading an `Image`:
```cpp
		Ref<Image> img;
		img.instance();
		Error err = img->load(item.icon.replace_first("res://", item.path + "/"));
		if (err == OK) {
			img->resize(default_icon->get_width(), default_icon->get_height(), Image::INTERPOLATE_LANCZOS);
			Ref<ImageTexture> it = memnew(ImageTexture);
			it->create_from_image(img);
			icon = it;
		}
```

**Comment by a3p169** [comment]
I just double checked and Godot's theme file, when asked to set by editor, does not use path in context to any project because theme file can be anywhere in the system. So a resource which is not in any project should not use 'res://' for its sub resources. Aren't sub resources embedded in '.res' or '.tres' file?

**Comment by Calinou** [comment]
> Aren't sub resources embedded in '.res' or '.tres' file?

They can be, but it's not a requirement.

**Comment by pycbouh** [comment]
The theme resource you've presented has an example of both: images are packed into it as a byte array, but fonts (DynamicFontData to be precise) is external:
```
[ext_resource path="res://fonts/opensans/OpenSans-Bold.ttf" type="DynamicFontData" id=1]
[ext_resource path="res://fonts/iosevka/iosevka-extended.ttf" type="DynamicFontData" id=2]
[ext_resource path="res://fonts/opensans/OpenSans-Regular.ttf" type="DynamicFontData" id=3]
```

**Comment by autofool** [comment]
@pycbouh Is there an update on this? Project Manager is still "fontless" in 3.3.3.stable. 
One off topic question since you are the one working on theme related stuff. There's this label in debugger bottom tab call "Stack Frames". It would be nice if  I could know which element is that? That's the only one I cant seem to change.

![20210923212337_SCREEN](https://user-images.githubusercontent.com/50569653/134541999-c11b1f0d-fc01-4734-a3e5-6b5be83d27df.png)

**Comment by pycbouh** [comment]
> Is there an update on this? Project Manager is still "fontless" in 3.3.3.stable.

We have no system for referencing resources globally (or relatively to each other for that matter), they can only be linked within a project with project-relative paths. So I don't think we can easily find a solution in the near future.

I've not looked into the fallback issue, which should be solvable. In the end, we should probably disable custom themes in the PM due to these problems. Or at least sanitize it to remove external resources somehow (so that colors, for example, are retained, but not fonts or icons).

> One off topic question since you are the one working on theme related stuff. There's this label in debugger bottom tab call "Stack Frames". It would be nice if I could know which element is that? That's the only one I am not able to change.

Isn't it just a header of a Tree control?

**Comment by autofool** [comment]
> We have no system for referencing resources globally (or relatively to each other for that matter), they can only be linked within a project with project-relative paths.
What about forcing of fonts as embedded resource in theme file?  
> In the end, we should probably disable custom themes in the PM due to these problems.
That's sounds good. A user doesn't spend too much time in Project Manager anyway. Most of the  session is spent in editor so having only editor affected by theme is reasonable.
> Isn't it just a header of a Tree control?
Unfortunately, current theme resource has no property to modify header of tree control. As you can see in bottom left corner, I have already customized whole tree control. Thanks!

**Comment by pycbouh** [comment]
> Unfortunately, current theme resource has no property to modify header of tree control. As you can see in bottom left corner, I have already customized whole tree control. Thanks!

I checked, and it's definitely the title row. It seems to be drawn using the following properties:

* title_button_color 
* title_button_font

* title_button_normal
* title_button_hover 
* title_button_pressed

The relevant code is here:

https://github.com/godotengine/godot/blob/3.3/scene/gui/tree.cpp#L3057-L3072

**Comment by autofool** [comment]
Got it! Thanks a bunch!

**Comment by autofool** [comment]
@pycbouh I just tried 3.4 beta and Project Manager has default fonts! It's perfect!
![20210923234316_WINDOW](https://user-images.githubusercontent.com/50569653/134561540-83c20b10-0da9-4da6-99f1-edbc88a09ca6.png)

**Comment by pycbouh** [comment]
Isn't project name still missing? Or is that an issue with your theme? 3.4 supports partial editor themes, so it should fill the gaps for missing stuff, but I'm not sure it's the complete cure.

**Comment by autofool** [comment]
Whoops! Totally missed that in excitement but that's an improvement for sure. Only titles are missing cause they use Bold fonts? It was produced in master:

> This is how the bug manifests itself with the `master` branch with the `theme.tres` included in the MRP:
> 
> ![image](https://user-images.githubusercontent.com/180032/117458397-8e040180-af4a-11eb-8f59-9035e7f24ea1.png)
> 
> Note that fonts are handled differently in the `master` branch. This is why the default _project_ font seems to be used everywhere, and the bold font (used for project titles) is replaced with tofu instead.
> 
> In 3.3:
> 
> ![image](https://user-images.githubusercontent.com/180032/117458676-d3c0ca00-af4a-11eb-84c7-74990e97ffa9.png)
> 
> Only the bold font is missing. A custom theme could provide one by adding [a font named `bold` to the `EditorFonts` class](https://github.com/godotengine/godot/blob/64403ddcab9f1dca2408f0a412a22d899708bbb1/editor/project_manager.cpp#L2464), but I'm not sure if this is possible from the current theme editor.
> 
> **Edit:** It's possible, see below.

**Comment by autofool** [comment]
Do note that the in 3.3.stable; second shot has all the fonts except title but current 3.3.3.stable has no fonts at all.

**Comment by autofool** [comment]
This is 3.3.3 stable official build:

![20210923235159_SCREEN](https://user-images.githubusercontent.com/50569653/134562570-7cffb1e6-e08a-4fe8-93ec-39fc488f96d7.png)

**Comment by autofool** [comment]
Sadly, I can not confirm whether 3.4 fills the gap left by custom theme cause beta5 crashes when custom theme is applied. #51648 should fix it though. Then, I'll confirm again if default fonts work for project title and everything then I don't think we need more solid solution.  At least not any time soon. But if 3.4 fails to show fonts then I guess we can keep this issue reopened for a reminder.

**Comment by pycbouh** [comment]
https://github.com/godotengine/godot/pull/51648 is in beta5...

**Comment by autofool** [comment]
Whoops! Should I reopen #52530 then?

**Comment by pycbouh** [comment]
If you think it's the same crash, sure. But make sure to update the reproduction steps for 3.4. And if you can, attach the resource that causes the problem.

**Comment by autofool** [comment]
Yes but I'll reopen tomorrow. I'm commenting from phone app and I can't find the option.

**Comment by Cammymoop** [comment]
This was fixed by #51648
